controlplane/kubeadm/controllers/upgrade.go

```diff
@@ -64,38 +67,54 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, errors.Wrap(err, "failed to set role and role binding for kubeadm")
 	}
 
-	if err := workloadCluster.UpdateKubernetesVersionInKubeadmConfigMap(ctx, parsedVersion); err != nil {
+	configMapKey := ctrlclient.ObjectKey{Name: "kubeadm-config", Namespace: metav1.NamespaceSystem}
+	kubeadmConfig, err := workloadCluster.GetKubeadmConfig(ctx, configMapKey)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to fetch kubeadm config map")
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(kubeadmConfig.GetConfigMap(), r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := kubeadmConfig.UpdateKubernetesVersion(parsedVersion); err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "failed to update the kubernetes version in the kubeadm config map")
 	}
 
 	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration != nil {
 		imageRepository := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository
-		if err := workloadCluster.UpdateImageRepositoryInKubeadmConfigMap(ctx, imageRepository); err != nil {
+		if err := kubeadmConfig.UpdateImageRepository(imageRepository); err != nil {
 			return ctrl.Result{}, errors.Wrap(err, "failed to update the image repository in the kubeadm config map")
 		}
 	}
 
 	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration != nil && kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.Local != nil {
 		meta := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.Local.ImageMeta
-		if err := workloadCluster.UpdateEtcdVersionInKubeadmConfigMap(ctx, meta.ImageRepository, meta.ImageTag); err != nil {
+		if changed, err := kubeadmConfig.UpdateEtcdMeta(meta.ImageRepository, meta.ImageTag); err != nil || !changed {
```

Comment:
The method under workload cluster was returning an error if there was no change. Hence added this check here to keep parity between the two interfaces.

Reply:
This is not blocking, but given that we are refactoring this part, I'm +1 to cleanup this return value because, as far as I can tell from this PR, it is not required anymore by the callers.
Generalizing a little bit, IMO it would be nice if all the Update methods could get a consistent signature, instead of having a few of them returning changed, and few others not